### PR TITLE
[WIP] Fix bug 1317230. Set custom homepage.

### PR DESCRIPTION
**Pull request by jotes**
@mathjazz could you do initial review and help me with that UI?
Currently I assign custom_homepage after first contribution (suggestion) -> If you think that doesn't makes a sense, just comment and I'll fix that.

### CODE REVIEW

File: pontoon/base/views.py [hunk 1]
```diff
@@ -56,10 +56,17 @@
...
+    user = request.user
+
+    if user.is_authenticated() and user.profile.custom_homepage:
+        return redirect('pontoon.teams.info', locale=user.profile.custom_homepage.code)
+
     project = Project.objects.get(id=1)
-    locale = utils.get_project_locale_from_request(
-        request, project.locales) or 'en-GB'
-    path = Resource.objects.filter(project=project, translatedresources__locale__code=locale)[0].path
+    locale = utils.get_project_locale_from_request(request, project.locales)
```
Comment (mathjazz) [inline pontoon/base/views.py]: Why did you remove the 'en-GB' piece?

File: pontoon/base/views.py [hunk 2]
```diff
@@ -660,6 +667,12 @@ def update_translation(request):
     user = request.user
     project = e.resource.project
 
+    def set_user_custom_homepage():
+        """Save custom homepage for user if it's not set"""
+        if not user.profile.custom_homepage:
+            user.profile.custom_homepage = l
+            user.profile.save()
+
```
Comment (mathjazz) [inline pontoon/base/views.py]: Saving translations needs to be blazing fast, so we should get rid of all non-essential operations.

It should be enough that users can set their custom homepage in settings. DB migration is a plus.

File: pontoon/base/views.py (hunk 1 above)
Comment (jotes) [inline pontoon/base/views.py]: Fixed, that was a remnant of the previous version.

File: pontoon/base/views.py (hunk 2 above)
Comment (jotes) [inline pontoon/base/views.py]: ok, removed.

File: pontoon/base/forms.py [hunk 3]
```diff
@@ -163,6 +163,15 @@ class Meta:
...
+    """
+    Form is responsible for saving custom home page.
+    """
+    class Meta:
+        model = UserProfile
+        fields = ('custom_homepage',)
+
+
 class UserLocalesSettings(forms.ModelForm):
     """
```
Comment (mathjazz) [inline pontoon/base/forms.py]: Not directly related to this PR, but I'd change the names of these forms:
`UserProfileForm` -> `UserFirstNameForm`
`UserCustomHomepageForm` -> this one is OK
`UserLocalesSettings` -> `UserLocalesOrder`

UserProfileForm also misses a comment: *Form is responsible for saving user's name.*

File: pontoon/contributors/views.py [hunk 4]
```diff
@@ -134,6 +134,20 @@ def save_user_name(request):
 
 
 @login_required(redirect_field_name='', login_url='/403')
+@require_POST
+@transaction.atomic
+def save_custom_homepage(request):
+    """Save user name."""
```
Comment (mathjazz) [inline pontoon/contributors/views.py]: Save custom homepage. :)

File: pontoon/contributors/templates/contributors/settings.html [hunk 5]
```diff
@@ -33,6 +34,8 @@ <h2 id="username">
   {% if user.translated_locales %}
     {{ Checkbox.checkbox('Make suggestions', class='force-suggestions', attribute='force_suggestions', is_enabled=user.profile.force_suggestions, title='Save suggestions instead of translations') }}
   {% endif %}
+  
+  {{ LocaleHomepage.select(user.profile.custom_homepage, all_locales) }}
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html]: Have you tried reusing [pontoon/teams/templates/teams/widgets/team_selector.html](#diff-a4ff7ce5b3f91c38b82a5e9fc30eb73c) here? It's basically the same selector, with additional "Default Pontoon Homepage" option. And a label.

File: pontoon/base/views.py [hunk 6]
```diff
@@ -754,7 +765,6 @@ def update_translation(request):
                     t.fuzzy = False
 
                     t.save()
-
                     return JsonResponse({
```
Comment (mathjazz) [inline pontoon/base/views.py]: Whoops.

File: pontoon/base/views.py [hunk 7]
```diff
@@ -731,7 +743,6 @@ def update_translation(request):
                     t.approved_date = now
 
                 t.save()
-
```
Comment (mathjazz) [inline pontoon/base/views.py]: Whoops.

**Comment by jotes** [comment]
@mathjazz I excluded locales_order from UserProfile form in admin because there's no nice widget for that field. I can fix that in the next PR if that's not urgent.

File: pontoon/base/models.py [hunk 8]
```diff
@@ -344,6 +344,9 @@ class UserProfile(models.Model):
     quality_checks = models.BooleanField(default=True)
     force_suggestions = models.BooleanField(default=False)
 
+    """Redirect user to a custom team page after login."""
```
Comment (mathjazz) [inline pontoon/base/models.py]: Used to redirect a user to a custom team page.

File: pontoon/base/views.py [hunk 9]
```diff
@@ -56,7 +56,15 @@
...
 def home(request):
     """Home view."""
+
+    user = request.user
     project = Project.objects.get(id=1)
+
+    # Redirect user to the selected home page or '/'. Because of circular redirects at home() this logic is moved to
+    # a separate view.
+    if user.is_authenticated() and user.profile.custom_homepage:
+        return redirect('pontoon.teams.info', locale=user.profile.custom_homepage.code)
```
Comment (mathjazz) [inline pontoon/base/views.py]: `'pontoon.teams.team'`

File: pontoon/base/views.py [hunk 10]
```diff
@@ -56,7 +56,15 @@
 
 def home(request):
     """Home view."""
+
+    user = request.user
     project = Project.objects.get(id=1)
+
+    # Redirect user to the selected home page or '/'. Because of circular redirects at home() this logic is moved to
+    # a separate view.
```
Comment (mathjazz) [inline pontoon/base/views.py]: You can remove the 2nd sentence now and make the comment shorter than 100 characters by doing that. :)

File: pontoon/base/views.py [hunk 11]
```diff
@@ -56,7 +56,15 @@
...
+    if user.is_authenticated() and user.profile.custom_homepage:
```
Comment (mathjazz) [inline pontoon/base/views.py]: How 'bout this - if the custom_homepage is not set, we [compute](https://irccloud.mozilla.com/pastebin/HthfU0wU/) it and set it?

File: pontoon/base/views.py (hunk 10 above)
Comment (jotes) [inline pontoon/base/views.py]: :D <3

**Comment by mathjazz** [comment]
@jotes This is now fully finished from my end. Selecting custom homepage from the settings page works. As such, I'd like to ask you to review my code.

**Comment by jotes** [comment]
@mathjazz lgfm -> is there anything left to do on my side? Can You proceed with the review?

**Comment by jotes** [comment]
I think that @MikkCZ could look at this -> We could connect this new feature with `pontoon-tools` (that's my humble suggestion).

**Comment by MikkCZ** [comment]
Thanks for notifying @jotes. Pontoon Tools can definitely use this as a default rather than guessing from the browser (but I will keep the option to override it in place for now). I will be AFK until the next week, so I will find a moment then.
